akka-http/src/main/scala/akka/http/scaladsl/coding/Deflate.scala

```diff
@@ -86,74 +90,68 @@ private[http] object DeflateCompressor {
   }
 }
 
-class DeflateDecompressor(maxBytesPerChunk: Int = Decoder.MaxBytesPerChunkDefault) extends DeflateDecompressorBase(maxBytesPerChunk) {
-
-  override def createLogic(attr: Attributes) = new DecompressorParsingLogic {
-    var _inflater: Inflater = _
-    override def inflater: Inflater = if (_inflater != null) {
-      _inflater
-    } else {
-      throw new IllegalStateException("Inflater used before the header was analysed for wrapping")
-    }
+/** Internal API */
+@InternalApi
+private[coding] class DeflateDecompressor(maxBytesPerChunk: Int = Decoder.MaxBytesPerChunkDefault) extends DeflateDecompressorBase(maxBytesPerChunk) {
 
-    override val inflateState = new Inflate(true) {
+  override def createLogic(attr: Attributes) = new ParsingLogic {
+    /** Step that probes if the deflate stream contains a zlib wrapper */
+    case object ProbeWrapping extends ParseStep[ByteString] {
       override def onTruncation(): Unit = completeStage()
 
       override def parse(reader: ByteStringParser.ByteReader): ParseResult[ByteString] = {
-        val data = reader.remainingData
-        if (_inflater == null) {
-          _inflater = examineAndBuildInflater(data)
-        }
-        super.parse(new ByteStringParser.ByteReader(reader.takeAll()))
+        val inflater = examineAndBuildInflater(reader.remainingData)
+        ParseResult(None, new Inflate(inflater, noPostProcessing = true, ProbeWrapping))
       }
     }
 
     /**
-     * We examine the head byte of the incoming data stream, looking for the standard 0x8 in the first 4 bits of the  header
+     * We examine the head byte of the incoming data stream, looking for the standard 0x8 in the first 4 bits of the header
      * that marks a byte stream as using a wrapped deflate stream. If we don't find that header, we fall back to the no-wrap
-     * inflater
+     * inflater.
+     *
+     * Wrapped deflate streams usually start with the first byte being 0x78 == b01111000. The lower three bits == 000 would mean an
+     * uncompressed stream for an unwrapped deflate stream. However, in that case the remaining bits are undefined and
+     * would likely be set to zero as well. Therefore, checking the lower 4 bits to be b1000 seems to be sufficient to
+     * decide that a stream is wrapped.
+     *
      * More details on the structure of wrap and no-wrap headers of Deflate can be found at:
      * https://www.ietf.org/rfc/rfc1950.txt (wrap) and
      * https://www.ietf.org/rfc/rfc1951.txt (no-wrap)
      */
-    private def examineAndBuildInflater(bytes: ByteString): Inflater =
-      bytes.headOption.filter(b ⇒ (b & 0x0F) != 8)
-        .fold(new Inflater())(_ ⇒ new Inflater(true))
-
-    override def afterInflate = inflateState
-    override def afterBytesRead(buffer: Array[Byte], offset: Int, length: Int): Unit = {}
+    private def examineAndBuildInflater(bytes: ByteString): Inflater = {
+      val wrapped = (bytes.head & 0x0F) == 0x08
```

Comment:
The old one took into account that the `ByteString` could be empty, but now we are sure it isn't?

Reply:
Yes, exactly. It's an invariant of `ParseStep.parse` that the byte reader will contain at least a single byte to read.